### PR TITLE
Modinreach 316

### PR DESCRIPTION
**Pull request by Arghya22**
MODINNREACH-316 ( https://issues.folio.org/browse/MODINREACH-316 )

## Purpose
Clean up "Virtual" Items and Associated Loan, Request, and Inventory Records in case of CANCEL_REQUEST

## Approach
1. Delete client method added for deletion of virtual record 
2. New service created which handles deletion of virtual record so that it can be handled from all the other services.
3. Test cases added accordingly 
4. Karate updated

API to test: 

```
curl --location --request PUT 'http://localhost:9130/inn-reach/d2ir/circ/cancelrequest/1068/d2ir' \
--header 'x-okapi-token: eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ0ZXN0LXVzZXIiLCJ0eXBlIjoibGVnYWN5LWFjY2VzcyIsInVzZXJfaWQiOiIwMDAwMDAwMC0xMTExLTU1NTUtOTk5OS05OTk5OTk5OTk5OTIiLCJpYXQiOjE2NjY4NDc4MTMsInRlbmFudCI6InRlc3RlbmFudDQxOTA5ODg1NDM4NDAxNTEzMzEifQ.Ji-I_n9t2FENDjMUkdi4Y2DAK-wTaOWzOC8DnHD_UsM' \
--header 'x-to-code: fli01' \
--header 'x-from-code: d2ir' \
--header 'x-d2ir-authorization: auth' \
--header 'x-okapi-tenant: testenant4190988543840151331' \
--header 'Content-Type: application/json' \
--data-raw '{
  "itemAgencyCode": "plgen",
  "patronId": "fpbg4dg3rffcdchjgf35apzcf4",
  "transactionTime": 1657026193,
  "patronAgencyCode": "fl2g1",
  "itemId": "it00000000085",
  "reason": "testing--",
  "reasonCode": 1
}'
```

### CODE REVIEW

File: src/main/java/org/folio/innreach/domain/service/impl/HoldingsServiceImpl.java [hunk 1]
```diff
@@ -34,6 +36,13 @@ public Optional<Holding> find(UUID holdingId) {
     return holdingsStorageClient.findHolding(holdingId);
   }
 
+  @Override
+  public void delete(UUID holdingId) {
+      holdingsStorageClient.findHolding(holdingId)
+        .ifPresentOrElse(holding -> holdingsStorageClient.deleteHolding(holdingId)
+          ,()->log.info("Holding not found with holdingId:{}",holdingId));
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/HoldingsServiceImpl.java]: Think you want spaces after the commas. Also a space after the `->`. Maybe put the comma that is at the beginning of the line at the end of the line above.

File: src/main/java/org/folio/innreach/domain/service/impl/InstanceServiceImpl.java [hunk 2]
```diff
@@ -39,6 +40,13 @@ public Optional<InventoryInstanceDTO> find(UUID uuid) {
     return inventoryClient.findInstance(uuid);
   }
 
+  @Override
+  public void delete(UUID instanceId) {
+    inventoryClient.findInstance(instanceId)
+      .ifPresentOrElse(instance-> inventoryClient.deleteInstance(instanceId)
+        ,()->log.info("Instance not found with instanceId:{}",instanceId));
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/InstanceServiceImpl.java]: Same comment as above.

File: src/main/java/org/folio/innreach/domain/service/impl/ItemServiceImpl.java [hunk 3]
```diff
@@ -37,6 +38,13 @@ public Optional<InventoryItemDTO> find(UUID itemId) {
     return inventoryClient.findItem(itemId);
   }
 
+  @Override
+  public void delete(UUID itemId) {
+      inventoryClient.findItem(itemId).
+        ifPresentOrElse(p -> inventoryClient.deleteItem(itemId),
+          () -> log.info("No item found with itemId:{}",itemId));
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/ItemServiceImpl.java]: Need space after comma

File: src/main/java/org/folio/innreach/domain/service/impl/LoanServiceImpl.java [hunk 4]
```diff
@@ -46,6 +46,13 @@ public Optional<LoanDTO> find(UUID loanId) {
     return circulationClient.findLoan(loanId);
   }
 
+  @Override
+  public void delete(UUID loanId) {
+    circulationClient.findLoan(loanId)
+      .ifPresentOrElse(loanDTO -> circulationClient.deleteLoan(loanId)
+        ,()->log.info("Loan not found with loanId:{}",loanId));
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/LoanServiceImpl.java]: Same as above

File: src/main/java/org/folio/innreach/domain/service/impl/RequestServiceImpl.java [hunk 5]
```diff
@@ -257,6 +257,14 @@ public RequestDTO findRequest(UUID requestId) {
       "No request found with id = " + requestId));
   }
 
+  @Override
+  public void deleteRequest(UUID requestId) {
+    circulationClient.findRequest(requestId)
+      .ifPresentOrElse(requestDTO -> circulationClient.deleteRequest(requestId)
+        ,()-> log.info("Request not present with requestid:"+requestId));
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/RequestServiceImpl.java]: Same

File: src/main/java/org/folio/innreach/domain/service/impl/VirtualRecordServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+@RequiredArgsConstructor
+public class VirtualRecordServiceImpl implements VirtualRecordService {
+
+  private final ItemService itemService;
+  private final HoldingsService holdingsService;
+  private final InstanceService instanceService;
+  private final LoanService loanService;
+  @Override
+  public void deleteVirtualRecords(UUID folioItemId, UUID folioHoldingId, UUID folioInstanceId, UUID folioLoanId) {
+    log.debug("deleteVirtualRecords :: parameters folioItemId:{} folioHoldingId:{} folioInstanceId:{} folioLoanId:{}",folioItemId,folioHoldingId,folioInstanceId,folioLoanId);
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/domain/service/impl/VirtualRecordServiceImpl.java]: Line is a bit long and needs spaces. Maybe break after the formatted string and put the params on a new line with spaces between them.

File: src/test/java/org/folio/innreach/controller/d2ir/CirculationApiTest.java [hunk 7]
```diff
@@ -241,10 +244,15 @@ void patronHold_cancelRequest() throws Exception {
...
     stubGet(format("%s/%s", ITEMS_URL, PRE_POPULATED_ITEM_ID), "inventory/item-response.json");
     stubPut(format("%s/%s", ITEMS_URL, PRE_POPULATED_ITEM_ID));
+    stubDelete(format("%s/%s", ITEMS_URL, PRE_POPULATED_ITEM_ID));
     stubGet(format("%s/%s", HOLDINGS_URL, NEW_HOLDING_ID), "inventory-storage/holding-response.json");
     stubPut(format("%s/%s", HOLDINGS_URL, NEW_HOLDING_ID));
+    stubDelete(format("%s/%s", HOLDINGS_URL, FOLIO_HOLDING_ID));
     stubGet(format("%s/%s", REQUESTS_URL, PRE_POPULATED_REQUEST_ID), "circulation/item-request-response.json");
     stubPut(format("%s/%s", REQUESTS_URL, PRE_POPULATED_REQUEST_ID));
+    stubDelete(format("%s/%s", INSTANCES_URL, FOLIO_INSTANCE_ID));
+    stubGet(LOAN_URL,"circulation/loan-response.json");
```
Comment (steveellis) [inline src/test/java/org/folio/innreach/controller/d2ir/CirculationApiTest.java]: Space after comma

File: src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java [hunk 8]
```diff
@@ -50,6 +50,7 @@
 import java.util.UUID;
 
 import org.apache.commons.lang3.ObjectUtils;
+import org.folio.innreach.domain.service.*;
```
Comment (steveellis) [inline src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java]: Usually wildcarding imports is discouraged.

File: src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java [hunk 9]
```diff
@@ -1132,6 +1132,7 @@ void processCancelRequest() {
       PRE_POPULATED_TRACKING1_ID, PRE_POPULATED_CENTRAL_CODE);
 
     verify(requestService).cancelRequest(anyString(), any(UUID.class), any(UUID.class), anyString());
+    verify(virtualRecordService).deleteVirtualRecords(any(UUID.class),any(UUID.class),any(UUID.class),any(UUID.class));
```
Comment (steveellis) [inline src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java]: Need spaces after commas for params.

File: src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java [hunk 10]
```diff
@@ -1122,6 +1121,7 @@ void testClaimsItemReturned_UnknownDate() {
   })
   void processCancelRequest() {
     doNothing().when(requestService).cancelRequest(anyString(), any(UUID.class), any(UUID.class), anyString());
+    doNothing().when(virtualRecordService).deleteVirtualRecords(any(UUID.class),any(UUID.class),any(UUID.class),any(UUID.class));
```
Comment (steveellis) [inline src/test/java/org/folio/innreach/controller/d2ir/InnReachCirculationControllerTest.java]: Spaces after commas

File: src/main/java/org/folio/innreach/domain/service/impl/HoldingsServiceImpl.java (hunk 1 above)
Comment (Arghya22) [reply src/main/java/org/folio/innreach/domain/service/impl/HoldingsServiceImpl.java]: done

File: src/main/java/org/folio/innreach/domain/service/impl/LoanServiceImpl.java (hunk 4 above)
Comment (steveellis) [reply src/main/java/org/folio/innreach/domain/service/impl/LoanServiceImpl.java]: Think this should be reformatted to follow the change you made above to be:
`() -> log.info("Loan not found with loanId: {}", loanId));`

Notice how I'm adding spaces.

File: src/main/java/org/folio/innreach/domain/service/impl/RequestServiceImpl.java (hunk 5 above)
Comment (steveellis) [reply src/main/java/org/folio/innreach/domain/service/impl/RequestServiceImpl.java]: Same as above. Please place the comma on the previous line on all of these, and add spaces around the `->` and the `+`.

File: src/main/java/org/folio/innreach/domain/service/impl/LoanServiceImpl.java (hunk 4 above)
Comment (steveellis) [reply src/main/java/org/folio/innreach/domain/service/impl/LoanServiceImpl.java]: Also move the comma to the previous line to be the same as line 45 above.

File: src/main/java/org/folio/innreach/domain/service/impl/VirtualRecordServiceImpl.java (hunk 6 above)
Comment (steveellis) [reply src/main/java/org/folio/innreach/domain/service/impl/VirtualRecordServiceImpl.java]: This line would be more readable as (notice the new line/break). Also add spaces between params:
```
log.debug("deleteVirtualRecords :: parameters folioItemId: {} folioHoldingId: {} folioInstanceId: {} folioLoanId: {}",
  folioItemId, folioHoldingId, folioInstanceId, folioLoanId);
```

**Review by steveellis — APPROVED** [review]

**Review by gurleenkaurbp — APPROVED** [review]

**Review by nirmalr-shah — APPROVED** [review]

**Pull request merged**